dashboard/views.py

```diff
@@ -143,6 +145,86 @@ def post(self, request):
         )
 
 
+class UserExamEnrollment(APIView):
+    """
+    Create an exam enrollment for the user in a given exam run identified by course_id.
+    """
+    authentication_classes = (
+        authentication.SessionAuthentication,
+        authentication.TokenAuthentication,
+    )
+    permission_classes = (permissions.IsAuthenticated, )
+
+    def post(self, request):
+        """
+        enrolls the user in a exam run in edx
+        """
+        edx_exam_course_id = request.data.get('exam_course_id')
+
+        if edx_exam_course_id is None:
+            raise ValidationError('exam course id missing in the request')
+
+        exam_run = ExamRun.objects.get(edx_exam_course_key=edx_exam_course_id)
+        if not ExamAuthorization.objects.filter(user=request.user, status=ExamAuthorization.STATUS_SUCCESS).exists():
+            raise ValidationError('user is not authorized for exam run')
+        provider = exam_run.course.first_unexpired_run().courseware_backend
+        url = urljoin(COURSEWARE_BACKEND_URL[provider], '/courses/{}/'.format(edx_exam_course_id))
+        # get the credentials for the current user for edX
+        user_social = get_social_auth(request.user, provider)
+        try:
+            utils.refresh_user_token(user_social)
+        except utils.InvalidCredentialStored as exc:
+            log.error(
+                "Error while refreshing credentials for user %s",
+                request.user.username,
+            )
+            return Response(
+                status=exc.http_status_code,
+                data={'error': str(exc)}
+            )
+
+        # create an instance of the client to query edX
+        edx_client = EdxApi(user_social.extra_data, COURSEWARE_BACKEND_URL[provider])
+        enrollments = edx_client.enrollments.get_student_enrollments()
+        all_enrolled_course_ids = enrollments.get_enrolled_course_ids()
+        # if user already enrolled in this exam course
+        if ExamRun.objects.filter(edx_exam_course_key__in=all_enrolled_course_ids).exists():
+            return Response({'url': url})
```

Comment:
I think this might cause problems if the user was enrolled in a previous exam run but is trying to take it a second time and isn't enrolled yet. We'd end up not enrolling them in the newer run but linking them to it.

Reply:
👍